x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportOpenJobAction.java

```diff
@@ -890,9 +894,15 @@ protected void nodeOperation(AllocatedPersistentTask task, OpenJobAction.JobPara
                 return;
             }
 
+            String jobId = jobTask.getJobId();
             autodetectProcessManager.openJob(jobTask, e2 -> {
                 if (e2 == null) {
-                    task.markAsCompleted();
+                    FinalizeJobExecutionAction.Request finalizeRequest = new FinalizeJobExecutionAction.Request(new String[]{jobId});
+                    executeAsyncWithOrigin(client, ML_ORIGIN, FinalizeJobExecutionAction.INSTANCE, finalizeRequest,
+                            ActionListener.wrap(
+                                    response -> task.markAsCompleted(),
+                                    e -> logger.error("error finalizing job [" + jobId + "]", e)
+                            ));
```

Comment:
Why did we move this from the close action to the open one?

Reply:
Because. 

`autodetectProcessManager.openJob` takes a lambda which is called when the job is closed, that lambda has a generically useless name `handler` and as you can see calls a method on `jobTask` which we pass to the method. The same `JobTask` is also passed to the `close` and `kill` methods which eventually call the lambda parameter we pass here which calls a method on the `jobTask` we pass in `close` & `kill`. Basically we don't need the lambda parameter and it makes it very difficult to find the bit of code called on job close as it is set here in the open method. 

`task.markAsCompleted()` causes a clusterstate update at which point the migrator sees that the job is closed and is eligible for migration. The `FinalizeJobExecutionAction` call needs to happen before to prevent a race condition between ` FinalizeJobExecutionAction` updating the job config and the migrator reading the config from the clusterstate. This happens to be the simplest place to make that change, it's not a natural home but it simplifies the logic doing it here.